[SPARK-20597][SQL][SS][WIP] KafkaSourceProvider falls back on path as synonym for topic

Nimfadora opened the pull request:
## What changes were proposed in this pull request?

**KafkaSourceProvider** supports 'topic' option that sets target Kafka topic.
**KafkaSourceProvider** can use 'topic' column to assign rows to Kafka topics for writing.

It is proposed to treat 'path' option in start(path: String) and save(path: String) as alternative to setting 'topic' option. Path would designate the target topic when 'topic' option is not specified and will override 'topic' column value for topic.

Currently if topic in 'path' option is specified it overrides topic column and is overridden by 'topic' option (if present). IMHO this relation seems more straightforward than that specified in the ticket ('path' is the least precedence option).

for streaming it will look as follows:
```
df.writeStream
  .format("kafka")
  .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
  .start("topic")
```
for batch it will look as follows:
```
df.write
  .format("kafka")
  .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
  .save("topic")
```

## How was this patch tested?
Added and extended unit tests, local build&test

Nimfadora commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

Nimfadora commented:
@jaceklaskowski could you please look at this patch and give your opinion on question asked in TODO section? Thank you

gaborgsomogyi commented:
If something can be configured multiple places and the value differs I would throw exception. If 10 person asked pretty sure several ways will be stated `natural`.

Nimfadora commented:
@gaborgsomogyi thank you for your tip, exception will be added. Docs also will be added as stated in the PR description along with reply on the review.

Nimfad0ra committed:
[SPARK-20597][SQL][SS] KafkaSourceProvider falls back on path as synonym for topic

Nimfadora commented:
added docs and error handling when topics do not match

jaceklaskowski commented on docs/structured-streaming-kafka-integration.md:
```diff
@@ -499,20 +522,28 @@ StreamingQuery ds = df
   .start();
 
+// Write key-value data from a DataFrame to Kafka using the topic specified in the path option
+StreamingQuery ds = df
```
nit: The name `ds` is not consistent with the type which is `StreamingQuery`. Please use `sq` or `q` instead.

jaceklaskowski commented on docs/structured-streaming-kafka-integration.md:
```diff
@@ -499,20 +522,28 @@ StreamingQuery ds = df
 
+// Write key-value data from a DataFrame to Kafka using the topic specified in the path option
+StreamingQuery ds = df
+  .selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
```
nit: Since you added it I'd change it to `.select($"key" cast "string", $"value" cast "string")` even if it's the only place in the doc. It's simply more type-safe.

Nimfadora replied:
would it be better to change it in all places in this doc file if we anyway touch it?

gaborgsomogyi replied:
With the latest change don't fully understand the direction. Now half of the doc uses `selectExpr ` and the other part `select`. Still questioning the need, what is the benefit? If the `selectExpr` executed and properly formed not more unstable than the simple `select` in the doc right?

Nimfadora replied:
select is used for scala, as we cannot use same syntax for Java and Python ($), but we can use 
`select(col("key").cast("string"))` for all three languages if we really want it to be typesafe and looks the same way.

gaborgsomogyi replied:
I see the intention now, thanks for the explanation. I personally would leave as it is but not telling  the direction is wrong. Let's decide by a committer who potentially picks this up.

Nimfadora replied:
@jaceklaskowski what do you think: should we revert the changes here or use `select(col("key").cast("string"))` in all places to provide type safety?

jaceklaskowski commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala:
```diff
@@ -194,6 +194,19 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
+    (topicOption, pathOption) match {
+      case (Some(t), Some(p)) if t != p => throw new IllegalArgumentException("'topic' and 'path'"
+        + " options should match if both defined: '" + t + "' should match '" + p + "'")
+      case (Some(t), _) => Some(t)
```
I'd use `topicOption` instead.

jaceklaskowski commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala:
```diff
@@ -194,6 +194,19 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
+      case (Some(t), Some(p)) if t != p => throw new IllegalArgumentException("'topic' and 'path'"
+        + " options should match if both defined: '" + t + "' should match '" + p + "'")
+      case (Some(t), _) => Some(t)
+      case (None, Some(p)) => Option(p)
```
I'd use `pathOption` instead.

jaceklaskowski commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala:
```diff
@@ -194,6 +194,19 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
+        + " options should match if both defined: '" + t + "' should match '" + p + "'")
+      case (Some(t), _) => Some(t)
+      case (None, Some(p)) => Option(p)
+      case _ => Option.empty
```
`None`?

jaceklaskowski reviewed (commented):
I'd like to have another example that shows what happens when all the possible options are in use.

gaborgsomogyi commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -160,76 +204,93 @@ class KafkaSinkSuite extends StreamTest with SharedSQLContext with KafkaTest {
   }
 
-  test("streaming - write aggregation w/o topic field, with topic option") {
+  /**
```
Not sure this description needed. The code should reflect the intention and add comment if it's tricky (I know it's from the legacy code).
Such comments can deviate from code (where is 'bar' for example) and can cause confusion.

gaborgsomogyi commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -160,76 +204,93 @@ class KafkaSinkSuite extends StreamTest with SharedSQLContext with KafkaTest {
-    } finally {
-      writer.stop()
-    }
+  test("streaming - write aggregation with path option") {
```
Some tests are named `streaming - write aggregation...` and others `streaming - aggregation...`. Is it intentional?

Nimfadora replied:
no, just haven't noticed that, thank you.

gaborgsomogyi reviewed (commented):
Thanks for the efforts! I've added my questions/thoughts.

Nimfad0ra committed:
[SPARK-20597][SQL][SS] add docs and refactor

Nimfadora commented on docs/structured-streaming-kafka-integration.md:
````diff
@@ -457,8 +463,17 @@ The following configurations are optional:
+  <td>streaming and batch</td>
   <td>Sets the topic that all rows will be written to in Kafka. This option overrides any
-  topic column that may exist in the data.</td>
+  topic column that may exist in the data and is overridden by ```topic``` option.
````
How about the use-case when you want to have 'topic' column in your data with different topic values in it (ex: "win", "click","visit"), but want to write that data to a single topic, so you override topic setting by topic option?

gaborgsomogyi replied:
Not yet understand the use-case, can you show the smallest example?

Nimfadora replied:
I rethought the situation and understood that you have the point, such situation is impossible, as we can only have "key", "value" and "topic" columns in the DataFrame that is written to Kafka.

Nimfadora replied:
I agree with you, that all three of them should be checked. However, now we have the validation being  splitted and duplicated between KafkaSourceProvider#validateQuery, KafkaWriteTask#createProjection and KafkaSourceProvider#resolveTopic. We can refactor this moment and move this validations to one place, or just leave as is and add topic column and topic/path option comparison validation to KafkaSourceProvider#validateQuery. The fist is more complicated and error-prone, but will result in more readable code. On the other hand, second solution will not require so much code to be rewritten. Which way do you think is right?

gaborgsomogyi replied:
I agree the first solution is more complicated but not fully understand why is it error-prone? Having checks in 2 different places which are belonging together is the error-prone solution. Of course good unit test coverage is required.

Nimfadora replied:
@gaborgsomogyi I agree with you, that first solution seems to be right way to go. I hope that I will not miss any place where it is used, while writing unit tests. My concern is about checking the topic column value. We cannot move the projection out of KafkaWriteTask, but if we move all the validation inside, than it would no longer be fail-fast: in current implementation we are checking all that we can before RDD creation. Although, I do not really understand why checks are duplicated between [KafkaWriteTask#createProjection](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala#L101) and [KafkaWriter#validateQuery](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala#L45)? As I can see validateQuery is always called before createProjection. @gaborgsomogyi @jaceklaskowski do you know any reason for this duplication?
![image](https://user-images.githubusercontent.com/10544767/54438390-a059ca80-4747-11e9-8416-13a16009e2bd.png)

The solution I propose: 
- remove duplicated checks from KafkaWriteTask#createProjection
- move check for topic and path option match to KafkaWriter#validateQuery
- add topic option and topic attribute match validation to KafkaWriteTask

Nimfadora replied:
KafkaWriteTask and KafkaWriter are not public interfaces, and in all places where [KafkaRowWriter#createProjection](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala#L101) is called, it is preceded by [KafkaWriter#validateQuery](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala#L45). I do not understand the purpose of the double check between these methods? It was introduced in #17043, @tcondie @tdas could you please explain the purpose of validation duplication? 

![image](https://user-images.githubusercontent.com/10544767/54684731-27cb8300-4b26-11e9-80de-d6c7bf2f3755.png)
@gaborgsomogyi the reason, why first solution is error-prone, is that I don't actually understand the reason of the decisions taken. When I removed the validation from `KafkaRowWriter#createProjection` nothing failed.

And the second concern is as follows: in the topic column we can specify different values and our messages will be sent to different topics based on the value in particular row. When the topic option is specified it is applied for all messages, even if the topic column is present and has different values. So, I think that these two options have different purposes of use and do not intersect, so checking them for equality is just redundant. I can imagine only one situation when I will specify both column and option: when I want to use column approach with different values in production, but for testing purposes I specify topic option so that all messages come to one topic for a while.

And I think that is better to leave the [KafkaSourceProvider#resolveTopic](https://github.com/apache/spark/pull/23791/files#diff-eeac5bdf3a1ecd7b9f8aaf10fff37f05R197) function where it is: 
we should not only validate that topic path and option match when both are defined but also trim and decide on which value to take (as one of them or both may be null). If we move method to `KafkaWriter` we will have to call this method from several other classes and pass additional path param and now all that logic is incapsulated inside `KafkaSourceProvider`.

I introduced the approach described here in my last commit.

gaborgsomogyi replied:
I've taken a deeper look and as you've mentioned it's a valid use-case where different topics are written because the topic column value differs. As a result the originally added fallback solution can be the way.

Nimfadora commented on docs/structured-streaming-kafka-integration.md:
```diff
@@ -539,35 +578,43 @@ ds = df \
+  .format("kafka")
+  .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+  .option("topic", "topic1")
+  .save()
```
Yes, missed that, thank you. Noticed only today, while addressing jaceklaskowski comments.

Nimfadora replied:
@gaborgsomogyi I don't quite understand where a java version  should be specified, could you explain, please?

gaborgsomogyi replied:
The java example in the java tab.

Nimfadora commented on docs/structured-streaming-kafka-integration.md:
````diff
@@ -457,8 +463,17 @@ The following configurations are optional:
   <td>none</td>
   <td>streaming and batch</td>
+  <td>Sets the topic that all rows will be written to in Kafka. This option overrides
+  ```path``` option and any topic column that may exist in the data.</td>
````
agree, already changed

Nimfad0ra committed:
[SPARK-20597][SQL][SS] minor refactoring

gaborgsomogyi commented:
Seems like another guy tagged and just seen the last comment. Could you please resolve the conversations where you think it's solved? It would help a lot.

Nimfadora commented:
@gaborgsomogyi I have resolved all addressed comments, and have several questions to you: ([1](https://github.com/apache/spark/pull/23791#discussion_r260283201), [2](https://github.com/apache/spark/pull/23791#discussion_r261521069)), thank you!

Nimfad0ra committed:
merge with master

Nimfad0ra committed:
[SPARK-20597][SQL][SS] remove duplications and refactor

Nimfadora commented:
@jaceklaskowski @gaborgsomogyi kindly ask you to review my answers once more. Thank you!

Nimfadora commented:
@jaceklaskowski kindly ping

Nimfadora commented:
@jaceklaskowski @tcondie @tdas could you please look at these changes?

AmplabJenkins commented:
Can one of the admins verify this patch?
